File: Beecrowd_Parte3/Morse_2338.java
import java.util.Scanner;

public class Main {

    // Arrays para armazenar os códigos Morse e os caracteres correspondentes
    private static final String[] morseCodes = {
            "=.===", "===.=.=.=", "===.=.===.=", "===.=.=", "=", "=.=.===.=",
            "===.===.=", "=.=.=.=","=.=", "=.===.===.===", "===.=.===", "=.===.=.=",
            "===.===", "===.=", "===.===.===", "=.===.===.=", "===.===.=.===",
            "=.===.=", "=.=.=", "===", "=.=.===", "=.=.=.===", "=.===.===",
            "===.=.=.===", "===.=.===.===", "===.===.=.="};

    private static final char[] characters = {
            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
            'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', ' '};

    // Método para decodificar uma mensagem em código morse
    private static String decodeMorse(String morseCode) {
        StringBuilder sb = new StringBuilder();

        // Divide o código morse pelos espaços entre palavras (.......)
        String[] words = morseCode.split("\\.{7}");
        
        for (String word : words) {
            String[] symbols = word.split("\\.{3}");

            for (String symbol : symbols) {
                char decodedChar = decodeSymbol(symbol);
                sb.append(decodedChar);
            }
            sb.append(" "); // Adiciona espaço entre palavras
        }

        // Remove o último espaço entre palavras, se houver
        if (sb.length() > 0 && sb.charAt(sb.length() - 1) == ' ') {
            sb.deleteCharAt(sb.length() - 1);
        }

        return sb.toString();
    }

    // Método para decodificar um único símbolo de código morse
    private static char decodeSymbol(String symbol) {
        for (int i = 0; i < morseCodes.length; i++) {
            if (morseCodes[i].equals(symbol)) {
                return characters[i];
            }
        }
        return ' '; // Se não encontrar, retorna espaço
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Lê o número de casos de teste
        int t = scanner.nextInt();
        scanner.nextLine(); // Consume a quebra de linha após o número de casos de teste

        // Processa cada caso de teste
        for (int i = 0; i < t; i++) {
            String morseCode = scanner.nextLine().trim();

            // Decodifica o código morse
            String decodedMessage = decodeMorse(morseCode);

            // Imprime o resultado
            System.out.println(decodedMessage);
        }

        scanner.close();
    }
}

File: Beecrowd_Parte3/Morse_2338.c
#include <stdio.h>
#include <string.h>

// Função para decodificar um único código Morse
void decodeMorse(char *morse) {
    // Mapeamento de Morse para caracteres
    const char *morseTable[] = {"=.===", "===.=.=.=", "===.=.===.=", "===.=.=", "=", "=.=.===.=",
 "===.===.=", "=.=.=.=","=.=", "=.===.===.===", "===.=.===", "=.===.=.=",
 "===.===", "===.=", "===.===.===", "=.===.===.=", "===.===.=.===",
 "=.===.=", "=.=.=", "===", "=.=.===", "=.=.=.===", "=.===.===",
 "===.=.=.===", "===.=.===.===", "===.===.=.="}; // espaço

    // Variável para armazenar a mensagem decodificada
    char decoded[1001];
    decoded[0] = '\0'; // Inicializa a string vazia

    char *token;
    token = strtok(morse, "......."); // Separa por sete pontos para palavras

    while (token != NULL) {
        char *letter;
        letter = strtok(token, "..."); // Separa por três pontos para letras
        
        while (letter != NULL) {
            // Verifica se letter é um espaço em branco
            if (strcmp(letter, "") == 0) {
                strcat(decoded, " ");
            } else {
                // Procura letter no morseTable
                int found = 0;
                for (int i = 0; i < 27; i++) {
                    if (strcmp(letter, morseTable[i]) == 0) {
                        char ch = 'a' + i;
                        strncat(decoded, &ch, 1); // Adiciona a letra correspondente
                        found = 1;
                        break;
                    }
                }
                if (!found) {
                    strcat(decoded, "?"); // Caractere não reconhecido
                }
            }

            letter = strtok(NULL, "..."); // Próxima letra
        }
        
        strcat(decoded, " "); // Adiciona espaço entre palavras
        token = strtok(NULL, "......."); // Próxima palavra
    }

    // Remove o último espaço adicionado antes de imprimir
    decoded[strlen(decoded) - 1] = '\0';

    // Imprime a mensagem decodificada
    printf("%s\n", decoded);
}

int main() {
    int t;
    scanf("%d", &t); // Número de casos de teste
    
    getchar(); // Consume newline after t
    
    for (int i = 0; i < t; i++) {
        char morse[1001];
        fgets(morse, 1001, stdin); // Lê o código Morse
        morse[strcspn(morse, "\n")] = '\0'; // Remove o newline
        
        decodeMorse(morse); // Decodifica e imprime a mensagem
    }
    
    return 0;
}
